Tim/transparent session (#39)
TankerHQ/spec: encryption_formats.md

#### Usage

> This format was instead of v6 when the padding option was disabled. It is no longer used.

Reviewer: used?

Reviewer: ```suggestion
This format was used instead of v6 when the padding option was disabled. It is no longer used.
```

Reviewer: Still, the wording seems unclear to me.

Maybe we should reword like: `This format is no longer used. It has been superseded by ...`. What do you think?

Author: Updated this one and v6 with "superseded" wording, let me know if the wording looks reasonable

Reviewer: LGTM

#### Usage

> This was the simple resource encryption format. It is no longer used.

Reviewer: used to be?

Reviewer: ```suggestion
This used to be the simple resource encryption format. It is no longer used.
```

Reviewer: Or use the `superseded` wording suggested above.

Author: Updated, per above

| Resource ID / Seed | Fixed length    | 16 bytes          |
| Encrypted data     | Variable length | = clear data size |
| MAC                | Fixed length    | 16 bytes          |

This format has been designed for simple resource encryption within an encryption session.

#### Properties

Constant overhead: O = 49 bytes

#### Issues

None.

#### Usage

Used instead of format v10 when padding is disabled.

### Encryption format v10

#### Spec

| **Element**         | **Buffer type** | **Byte length**             |
|---------------------| --------------- |-----------------------------|
| Version number (10) | Fixed length    | 1 byte                      |
| Session ID          | Fixed length    | 16 bytes                    |
| Resource ID / Seed  | Fixed length    | 16 bytes                    |
| Encrypted data      | Variable length | = clear data size + padding |
| MAC                 | Fixed length    | 16 bytes                    |

This format has been designed for simple resource encryption within an encryption session.
It is the same as v9 but with the addition of padding. The padding algorithm and overhead is the same as v6.

#### Properties

> Constant overhead: O = 49 bytes

Reviewer 2: better to remind about the padding ovearhead here too

```suggestion
Constant overhead: O = 49 bytes

Padding overhead: O = padme_overhead(N + 1) bytes

*padme_overhead(N)* is detailed in the v6 section.
```

| Version number (9) | Fixed length    | 1 byte            |
| Session ID         | Fixed length    | 16 bytes          |
| Resource ID / Seed | Fixed length    | 16 bytes          |
| Encrypted data     | Variable length | = clear data size |
| MAC                | Fixed length    | 16 bytes          |

This format has been designed for simple resource encryption within an encryption session.

#### Properties

Constant overhead: O = 49 bytes

#### Issues

None.

#### Usage

Used instead of format v10 when padding is disabled.

### Encryption format v10

#### Spec

| **Element**         | **Buffer type** | **Byte length**             |
|---------------------| --------------- |-----------------------------|
| Version number (10) | Fixed length    | 1 byte                      |
| Session ID          | Fixed length    | 16 bytes                    |
| Resource ID / Seed  | Fixed length    | 16 bytes                    |
| Encrypted data      | Variable length | = clear data size + padding |
| MAC                 | Fixed length    | 16 bytes                    |

This format has been designed for simple resource encryption within an encryption session.

> It is the same as v8 but with the addition of padding. The padding algorithm and overhead is the same as v6.

Reviewer: v8 -> v9

Reviewer: ```suggestion
It is the same as v9 but with the addition of padding. The padding algorithm and overhead is the same as v6.
```

#### Properties

Constant overhead: O = 49 bytes

#### Issues

None.

#### Usage

The session used when encrypting with this resource is created automatically (transparent session).

> It is used by default for simple resource encryption, when not using a manually created session.

Reviewer: * Wording: `with this resource` -> `with this format`?
* Swap sentences for clarity?
* Mention padding?

Reviewer: ```suggestion
It is used by default for simple resource encryption, when not using a manually created session or disabling padding.
The session used when encrypting with this format is created automatically (transparent session).
```

Author: Applied (with slightly different wording for padding, let me know if that looks good)

| Encrypted chunk MAC  | Fixed length         | 16 bytes           |

All encrypted chunks have the same size (= maximum encrypted chunk size), except the last one which has always a shorter size.

#### Properties

Total clear data size: TCDS (variable)

Header overhead: HO = 37 bytes

Overhead per chunk: CO = 16 bytes

Encrypted chunk size: ECS = variable (default value: 1 MB)

Clear chunk size: CCS = ECS - CO (default value: 1 MB - 16 bytes)

Number of chunks: NC = ceil[ (TCDS + 1) / CCS ]

Total variable overhead: O = HO + CO * NC

When the total **clear** data size is a multiple of the maximum **clear** chunk size, an additional “empty” chunk is added at the end. This ensures that the encrypted data is not truncated by a malicious storage service. Corresponding formula:
TCDS % (ECS - CO) = 0  =>  +1 “empty” chunk of length CO = 16 bytes

The Resource ID is randomly generated once, and is also the seed used to derive the individual encryption key of this resource.

The IV used in encryption is derived from the session ID and the index of the chunk with the following formula: IV = H(session ID, index). The index is concatenated as an uint64 little endian and starts from 0.

> This design avoids using the same Key+IV combination multiple times if we need to support chunk rewriting, while also protecting against a malicious storage service that tries to reorder the chunks.

Reviewer 2: It does not protect in case of rewriting (we don't have a random seed be chunk anymore) 

```suggestion
This design avoids using the same Key+IV combination multiple times while also protecting against a malicious storage service that tries to reorder the chunks.
```

#### Properties

Total clear data size: TCDS (variable)

Header overhead: HO = 37 bytes

Overhead per chunk: CO = 20 bytes

Padding overhead: PO = *padme_overhead(TCDS)*

*padme_overhead(N)* is detailed in the v6 section.

Encrypted chunk size: ECS = variable (default value: 1 MB)

Clear chunk size: CCS = ECS - CO (default value: 1 MB - 20 bytes)

Number of chunks: NC = ceil[ (TCDS + PO) / CCS ]

Total variable overhead: O = H0 + PO + CO * NC

When the total **clear** data size + padding overhead is a multiple of the maximum **clear** chunk size, an additional “empty” chunk is added at the end. This ensures that the encrypted data is not truncated by a malicious storage service. Corresponding formula:
(TCDS + PO) % (CCS) = 0  =>  +1 “empty” chunk of length CO = 20 bytes

The Resource ID is randomly generated once, and is also the seed used to derive the individual encryption key of this resource.

The IV used in encryption is derived from the session ID and the index of the chunk with the following formula: IV = H(session ID, index). The index is concatenated as an uint64 little endian and starts from 0.

> This design avoids using the same Key+IV combination multiple times if we need to support chunk rewriting, while also protecting against a malicious storage service that tries to reorder the chunks.

Reviewer 2: DITTO

```suggestion
This design avoids using the same Key+IV combination multiple times while also protecting against a malicious storage service that tries to reorder the chunks.
```

Encrypted chunk size: ECS = variable (default value: 1 MB)

Clear chunk size: CCS = ECS - CO (default value: 1 MB - 16 bytes)

Number of chunks: NC = ceil[ (TCDS + 1) / CCS ]

Total variable overhead: O = HO + CO * NC

When the total **clear** data size is a multiple of the maximum **clear** chunk size, an additional “empty” chunk is added at the end. This ensures that the encrypted data is not truncated by a malicious storage service. Corresponding formula:
TCDS % (ECS - CO) = 0  =>  +1 “empty” chunk of length CO = 16 bytes

The Resource ID is randomly generated once, and is also the seed used to derive the individual encryption key of this resource.

The IV used in encryption is derived from the session ID and the index of the chunk with the following formula: IV = H(session ID, index). The index is concatenated as an uint64 little endian and starts from 0.

This design avoids using the same Key+IV combination multiple times, while also protecting against a malicious storage service that tries to reorder the chunks.

The Key is derived from the Resource ID and the key of the encryption session, and is the same for all chunks.

#### Issues

None.

#### Usage

> Used instead of format v12 when padding is disabled.

Reviewer 2: remove this (and change version number in the new v11 spec)

| Encrypted chunk MAC  | Fixed length         | 16 bytes           |

All encrypted chunks have the same size (= maximum encrypted chunk size), except the last one which has always a shorter size.

#### Properties

Total clear data size: TCDS (variable)

Header overhead: HO = 37 bytes

Overhead per chunk: CO = 16 bytes

Encrypted chunk size: ECS = variable (default value: 1 MB)

Clear chunk size: CCS = ECS - CO (default value: 1 MB - 16 bytes)

Number of chunks: NC = ceil[ (TCDS + 1) / CCS ]

Total variable overhead: O = HO + CO * NC

When the total **clear** data size is a multiple of the maximum **clear** chunk size, an additional “empty” chunk is added at the end. This ensures that the encrypted data is not truncated by a malicious storage service. Corresponding formula:
TCDS % (ECS - CO) = 0  =>  +1 “empty” chunk of length CO = 16 bytes

The Resource ID is randomly generated once, and is also the seed used to derive the individual encryption key of this resource.

The IV used in encryption is derived from the session ID and the index of the chunk with the following formula: IV = H(session ID, index). The index is concatenated as an uint64 little endian and starts from 0.

> This designs avoids using the same Key+IV combination multiple times if we need to support chunk rewriting, while also protecting against a malicious storage service that tries to reorder the chunks.

Reviewer: designs -> design

#### Properties

Total clear data size: TCDS (variable)

Header overhead: HO = 37 bytes

Overhead per chunk: CO = 20 bytes

Padding overhead: PO = *padme_overhead(TCDS)*

*padme_overhead(N)* is detailed in the v6 section.

Encrypted chunk size: ECS = variable (default value: 1 MB)

Clear chunk size: CCS = ECS - CO (default value: 1 MB - 20 bytes)

Number of chunks: NC = ceil[ (TCDS + PO) / CCS ]

Total variable overhead: O = H0 + PO + CO * NC

When the total **clear** data size + padding overhead is a multiple of the maximum **clear** chunk size, an additional “empty” chunk is added at the end. This ensures that the encrypted data is not truncated by a malicious storage service. Corresponding formula:
(TCDS + PO) % (CCS) = 0  =>  +1 “empty” chunk of length CO = 20 bytes

The Resource ID is randomly generated once, and is also the seed used to derive the individual encryption key of this resource.

The IV used in encryption is derived from the session ID and the index of the chunk with the following formula: IV = H(session ID, index). The index is concatenated as an uint64 little endian and starts from 0.

> This designs avoids using the same Key+IV combination multiple times if we need to support chunk rewriting, while also protecting against a malicious storage service that tries to reorder the chunks.

Reviewer: designs -> design

The Key is derived from the Resource ID and the key of the encryption session, and is the same for all chunks.

#### Usage

> Currently used to encrypt clear data with byte length >= 1 MB (e.g. big files) when *not* using manual encryption sessions.

Reviewer: or disabling padding?

Reviewer: ```suggestion
Currently used to encrypt clear data with byte length >= 1 MB (e.g. big files) when *not* using manual encryption sessions or disabling padding.
```

Author: Added (with slightly different wording)
